dmoj/judge.py

```diff
@@ -459,69 +491,132 @@ def _grade_cases(self) -> Generator[Tuple[IPC, tuple], None, None]:
+            return result
+
+        def skip_single(case: AbstractTestCase) -> Tuple[IPC, tuple]:
+            return IPC.RESULT, (case.batch, case.position, Result(case, result_flag=Result.SC))
+
+        def skip_batched_cases(batched_cases: Iterable[BatchedTestCase]) -> Generator[Tuple[IPC, tuple], None, None]:
+            for case in batched_cases:
+                yield skip_single(case)
+
+        def skip_toplevel(test: TopLevelCase) -> Generator[Tuple[IPC, tuple], None, None]:
+            if isinstance(test, Batch):
+                yield IPC.BATCH_BEGIN, (test.batch,)
+                yield from skip_batched_cases(test.batched_cases)
+                yield IPC.BATCH_END, (test.batch,)
             else:
-                flattened_cases.append((None, case))
-
-        case_number = 0
-        is_short_circuiting = False
-        is_short_circuiting_enabled = self.submission.short_circuit
-        passed_batches: Set[int] = set()
-        for batch_number, cases in groupby(flattened_cases, key=itemgetter(0)):
-            if batch_number:
-                yield IPC.BATCH_BEGIN, (batch_number,)
-
-                dependencies = batch_dependencies[batch_number - 1]  # List is zero-indexed
-                if passed_batches & dependencies != dependencies:
-                    is_short_circuiting = True
-
-            for _, case in cases:
-                case_number += 1
-
-                # Stop grading if we're short circuiting
-                if is_short_circuiting:
-                    result = Result(case, result_flag=Result.SC)
+                yield skip_single(test)
+
+        def skip_many_toplevel(tests: List[TopLevelCase]) -> Generator[Tuple[IPC, tuple], None, None]:
+            for test in tests:
+                yield from skip_toplevel(test)
+
+        def grade_batched_cases(batched_cases: List[BatchedTestCase]) -> Generator[Tuple[IPC, tuple], None, int]:
+            case_iter = iter(batched_cases)
+            for case in case_iter:
+                result = grade(case)
+                yield IPC.RESULT, (case.batch, case.position, result)
+
+                if result.result_flag & Result.WA:
+                    # Skip the rest.
+                    yield from skip_batched_cases(case_iter)
+                    return Result.WA
+
+            return Result.AC
+
+        class GradeManyResult(Enum):
+            OK = 0
+            SKIP_ALL = 1
+
+        def grade_many(
+            tests: List[TopLevelCase], should_run: Callable[[TopLevelCase, Set[int]], bool]
+        ) -> Generator[Tuple[IPC, tuple], None, Tuple[Set[int], GradeManyResult]]:
+            failed_tests: Set[int] = set()
+            test_iter = iter(tests)
+            for test in test_iter:
+                if not should_run(test, failed_tests):
+                    failed = True
+                    yield from skip_toplevel(test)
+                elif isinstance(test, Batch):
+                    yield IPC.BATCH_BEGIN, (test.batch,)
```

Comment:
Can we move these yields into `grade_batched_cases`? It's a little weird that the case results come from within it, but the batch begin/end markers don't.